org.eclipse.jdt.junit.core/src/org/eclipse/jdt/internal/junit/util/CoreTestSearchEngine.java

```diff
@@ -173,19 +176,39 @@ private static boolean hasJUnitJupiterTestAnnotation(IJavaProject project, int j
 				}
 				if (type != null) {
 					// check if we have the right JUnit JUpiter version
+					Version version = null;
 					String filename= type.getPath().lastSegment();
-					if ((filename.startsWith(junitBundlePrefix + "_") || filename.startsWith(junitBundlePrefix + "-")) && filename.endsWith(JAR_EXTENSION)) { //$NON-NLS-1$ //$NON-NLS-2$
+					boolean isJar = filename.endsWith(JAR_EXTENSION);
+					// Try matching the library name, this should be enough for many cases.
+					if (isJar && (filename.startsWith(junitBundlePrefix + "_") || filename.startsWith(junitBundlePrefix + "-"))) { //$NON-NLS-1$ //$NON-NLS-2$
 						String versionString = filename.substring(junitBundlePrefix.length() + 1, filename.length() - JAR_EXTENSION.length());
 						try {
-							Version version = Version.parseVersion(versionString);
-							if (version.getMajor() != junitMajorVersion) {
-								return false;
-							}
+							version = Version.parseVersion(versionString);
 						} catch (IllegalArgumentException e) {
-							// Some JUnit distributions use different naming, ignore those. See: https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/2665
-							JUnitCorePlugin.log(Status.error("Failed to parse determine JUnit version: " + filename, e)); //$NON-NLS-1$
+							/*
+							 * Some JUnit distributions use different naming, ignore those. See: https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/2665
+							 * We will try parsing a jar manifest below.
+							 */
 						}
 					}
+					if (isJar && version == null) {
+						IPath typePathWithoutProject= type.getPath().removeFirstSegments(1);
+						IPath absolutePath = project.getProject().getLocation().append(typePathWithoutProject);
```

Comment:
Sure that the jar is always located inside the project and not part of the dependency chain (and so located somewhere else or even links to external location)?

Javadoc says

> If this element is not included in an external library, the path returned is the full, absolute path to the underlying resource, relative to the workbench. If this element is included in an external library, the path returned is the absolute path to the archive or to the folder in the file system. 

I would first ask `IWorkspaceRoot.getFileForLocation(IPath)` and if there is nothing, would check if `Files.get(type.getPath().toOSString())` exists